### PR TITLE
[extractor/nos.nl] Add extractor

### LINKED ISSUE #4649: https://nos.nl/
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.08.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Netherlands

### Provide a description that is worded well enough to be understood

https://nos.nl/
Videos from this website (the Dutch BBC) don't work: Unsupported URL..

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.08 [3157158] (win32_exe)
[debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: avconv -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: avprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.08.08, Current version: 2022.08.08
yt-dlp is up to date (2022.08.08)
[debug] [generic] Extracting URL: https://nos.nl/nieuwsuur/artikel/2440353-verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen
[generic] 2440353-verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] 2440353-verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen: Extracting information
[debug] Looking for Brightcove embeds
[debug] Looking for embeds
ERROR: Unsupported URL: https://nos.nl/nieuwsuur/artikel/2440353-verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1441, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1517, in __extract_info
  File "yt_dlp\extractor\common.py", line 666, in extract
  File "yt_dlp\extractor\generic.py", line 3077, in _real_extract
yt_dlp.utils.UnsupportedError: Unsupported URL: https://nos.nl/nieuwsuur/artikel/2440353-verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen
```

### PR DESCRIPTION
**Pull request by HobbyistDev**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

ADD DESCRIPTION HERE
This PR add site-support for nos.nl, a site that serve news in Netherland 

Fixes #4649 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/nosnl.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    def _real_extract(self, url):
+        display_id = self._match_valid_url(url).group('display_id')
+        webpage = self._download_webpage(url, display_id)
+
+        nextjs_json = self._search_nextjs_data(webpage, display_id)['props']['pageProps']['data']
+        return self.playlist_result(
+            self._get_video_generator(nextjs_json, display_id),
+            str(nextjs_json['id']), nextjs_json.get('title') or self._html_search_meta(['title', 'og:title', 'twitter:title'], webpage),
+            nextjs_json.get('description') or self._html_search_meta(['description', 'twitter:description', 'og:description'], webpage),
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:69]:
```suggestion
            str(nextjs_json['id']),
            title=nextjs_json.get('title') or self._html_search_meta(['title', 'og:title', 'twitter:title'], webpage),
            description=nextjs_json.get('description') or self._html_search_meta(['description', 'twitter:description', 'og:description'], webpage),
```

File: yt_dlp/extractor/nosnl.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            tags=nextjs_json.get('keyword'), modified_timestamp=parse_iso8601(nextjs_json.get(nextjs_json.get('modifiedAt'))))
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:70]:
```suggestion
            tags=nextjs_json.get('keyword'),
            modified_timestamp=parse_iso8601(nextjs_json.get(nextjs_json.get('modifiedAt'))))
```
Why is there no tests for these?

File: yt_dlp/extractor/nosnl.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                'id': '2440789',
+                'title': 'Wekdienst 16/8: Groningse acties tien jaar na zware aardbeving • Femke Bol in actie op EK atletiek ',
+                'description': 'Nieuws, weer, verkeer: met dit overzicht begin je geïnformeerd aan de dag.',
+            },
+            'playlist_count': 2,
+        },
+    ]
+
+    def _get_video_generator(self, nextjs_json, display_id):
+        for item in nextjs_json.get('items'):
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:40]:
```suggestion
        for item in nextjs_json['items']:
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/nosnl.py (hunk 2 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/nosnl.py]: Sorry, that was a typo, fixed in 2a17ab7029e2e6df995304378e9f903cd24d389d

File: yt_dlp/extractor/nosnl.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                formats, subtitle = self._extract_m3u8_formats_and_subtitles(
+                    traverse_obj(item, ('source', 'url')), display_id, ext='mp4')
+                yield {
+                    'id': str(item['id']),
+                    'title': item.get('title'),
+                    'description': item.get('description'),
+                    'formats': formats,
+                    'subtitles': subtitle,
+                    'duration': parse_duration(item.get('duration')),
+                    'thumbnails': self._get_video_thumbnails(item.get('imagesByRatio'))
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:74]:
```suggestion
                    'thumbnails': [{
                        'url': traverse_obj(image, ('url', ...), get_all=False),
                        'width': image.get('width'),
                        'height': image.get('height')
                    } for image in traverse_obj(item, ('imagesByRatio', ...))],
```

File: yt_dlp/extractor/nosnl.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                }
+
+    def _real_extract(self, url):
+        display_id = self._match_valid_url(url).group('display_id')
+        webpage = self._download_webpage(url, display_id)
+
+        nextjs_json = self._search_nextjs_data(webpage, display_id)['props']['pageProps']['data']
+        return {
+            '_type': 'playlist',
+            'entries': self._get_video_generator(nextjs_json, display_id),
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:92]: Conventionally, the name `_entries` is used

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by HobbyistDev** [comment]
84c2904eb44013b1b0706bbcb8b3e1c8df616158 give error in test:
<details> <summary> test result </summary>

```
(python_proj2) [REDACTED]>python test/test_download.py TestDownload.test_NOSNLArticle_all
[debug] Loaded 1692 extractors 
[debug] [NOSNLArticle] Extracting URL: https://nos.nl/nieuwsuur/artikel/2440353-verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen 
[NOSNLArticle] verzakking-door-droogte-dreigt-tot-een-miljoen-kwetsbare-huizen: Downloading webpage 
[download] Downloading playlist: Verzakking door droogte dreigt: 'Tot een miljoen kwetsbare huizen' 
E
======================================================================
ERROR: test_NOSNLArticle_all (__main__.TestDownload):
----------------------------------------------------------------------
Traceback (most recent call last):
  File "[REDACTED]\test\test_download.py", line 278, in test_template
    getattr(self, test_name)()
  File "[REDACTED]\test\test_download.py", line 159, in test_template
    res_dict = ydl.extract_info(
  File "[REDACTED]\yt_dlp\YoutubeDL.py", line 1466, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
  File "[REDACTED]\yt_dlp\YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "[REDACTED]\yt_dlp\YoutubeDL.py", line 1574, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "[REDACTED]\yt_dlp\YoutubeDL.py", line 1702, in process_ie_result
    return self.__process_playlist(ie_result, download)
  File "[REDACTED]\yt_dlp\YoutubeDL.py", line 1769, in __process_playlist
    all_entries = PlaylistEntries(self, ie_result)
  File "[REDACTED]\yt_dlp\utils.py", line 2917, in __init__
    raise EntryNotInPlaylist('There are no entries')
yt_dlp.utils.EntryNotInPlaylist: There are no entries

----------------------------------------------------------------------
Ran 1 test in 1.865s

FAILED (errors=1)

```

File: yt_dlp/extractor/nosnl.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                'modified_timestamp': 1660625449,
+                'timestamp': 1660625449,
+                'upload_date': '20220816',
+                'thumbnail': 'https://cdn.nos.nl/image/2022/08/16/888178/1024x576a.jpg',
+            },
+            'playlist_count': 2,
+        }
+    ]
+
+    def _get_video_generator(self, nextjs_json, display_id):
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:52]:
```suggestion
    def _entries(self, nextjs_json, display_id):
```

File: yt_dlp/extractor/nosnl.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                }
+
+    def _real_extract(self, url):
+        display_id = self._match_valid_url(url).group('display_id')
+        webpage = self._download_webpage(url, display_id)
+
+        nextjs_json = self._search_nextjs_data(webpage, display_id)['props']['pageProps']['data']
+        return {
+            '_type': 'playlist',
+            '_entries': self._get_video_generator(nextjs_json, display_id),
```
Comment (pukkandan) [inline yt_dlp/extractor/nosnl.py:86]:
```suggestion
            'entries': self._entries(nextjs_json, display_id),
```

**Review by pukkandan — APPROVED** [review]

**Pull request merged**